File: src/types/grok.ts
export interface GrokQuota {
  monthly_limit?: number | null;
  used?: number | null;
  remaining?: number | null;
  usage_percent?: number | null;
  remaining_percent?: number | null;
  on_demand_cap?: number | null;
  on_demand_used?: number | null;
  prepaid_balance?: number | null;
  billing_period_start?: string | null;
  billing_period_end?: string | null;
  unlimited_or_free?: boolean | null;
  exhausted?: boolean | null;
  exhaust_reason?: string | null;
}

export interface GrokAccount {
  id: string;
  email: string;
  name?: string | null;
  first_name?: string | null;
  last_name?: string | null;
  user_id?: string | null;
  principal_id?: string | null;
  team_id?: string | null;
  profile_image_asset_id?: string | null;
  tier?: number | null;
  plan_type?: string | null;
  plan_label?: string | null;
  access_token: string;
  refresh_token?: string | null;
  scope?: string | null;
  expires_at?: number | null;
  expires_at_raw?: string | null;
  oidc_issuer?: string | null;
  oidc_client_id?: string | null;
  auth_entry_key?: string | null;
  auth_mode_raw?: string | null;
  create_time?: string | null;
  coding_data_retention_opt_out?: boolean | null;
  has_grok_code_access?: boolean | null;
  quota?: GrokQuota | null;
  usage_updated_at?: number | null;
  token_updated_at?: number | null;
  status?: string | null;
  status_reason?: string | null;
  requires_reauth?: boolean | null;
  reauth_reason?: string | null;
  quota_query_last_error?: string | null;
  quota_query_last_error_at?: number | null;
  subscription_query_last_success_at?: number | null;
  tags?: string[] | null;
  account_note?: string | null;
  created_at: number;
  last_used: number;
}

export interface GrokOAuthStartResponse {
  loginId: string;
  verificationUri: string;
  userCode?: string | null;
  expiresIn: number;
  intervalSeconds: number;
  callbackUrl?: string | null;
}

export function getGrokAccountDisplayEmail(account: GrokAccount): string {
  const email = account.email?.trim();
  if (email) return email;
  const name = account.name?.trim();
  if (name) return name;
  const first = account.first_name?.trim() || '';
  const last = account.last_name?.trim() || '';
  const combined = `${first} ${last}`.trim();
  if (combined) return combined;
  return account.id;
}

export function getGrokPlanBadge(account: GrokAccount): string {
  const raw = (account.plan_type || account.plan_label || '').trim().toUpperCase();
  if (!raw) {
    if (account.tier === 5) return 'HEAVY';
    if (account.tier != null && account.tier >= 3) return 'SUPERGROK';
    if (account.tier != null && account.tier <= 1) return 'FREE';
    return 'UNKNOWN';
  }
  if (raw.includes('HEAVY')) return 'HEAVY';
  if (raw.includes('LITE')) return 'LITE';
  if (raw.includes('SUPER')) return 'SUPERGROK';
  if (raw.includes('FREE')) return 'FREE';
  if (raw.includes('ENTERPRISE')) return 'ENTERPRISE';
  return raw.slice(0, 12);
}

export function getGrokPlanBadgeClass(account: GrokAccount): string {
  const badge = getGrokPlanBadge(account);
  if (badge === 'HEAVY' || badge === 'ENTERPRISE') return 'ultra';
  if (badge === 'SUPERGROK' || badge === 'LITE') return 'pro';
  if (badge === 'FREE') return 'free';
  return 'unknown';
}

export function getGrokUsage(account: GrokAccount): {
  totalPercentUsed: number | null;
  remainingPercent: number | null;
  inlineSuggestionsUsedPercent: number | null;
  chatMessagesUsedPercent: number | null;
  allowanceResetAt: number | null;
} {
  const q = account.quota;
  const used = q?.usage_percent ?? null;
  const remaining =
    q?.remaining_percent ??
    (typeof used === 'number' ? Math.max(0, 100 - used) : null);
  let resetAt: number | null = null;
  if (q?.billing_period_end) {
    const parsed = Date.parse(q.billing_period_end);
    if (!Number.isNaN(parsed)) resetAt = Math.floor(parsed / 1000);
  }
  return {
    totalPercentUsed: used,
    remainingPercent: remaining,
    inlineSuggestionsUsedPercent: used,
    chatMessagesUsedPercent: used,
    allowanceResetAt: resetAt,
  };
}

export function hasGrokQuotaData(account: GrokAccount): boolean {
  return account.quota != null;
}

export function formatGrokQuotaSummary(account: GrokAccount): string {
  const q = account.quota;
  if (!q) return '--';
  if (q.unlimited_or_free) return 'Unlimited / Free';
  if (typeof q.used === 'number' && typeof q.monthly_limit === 'number' && q.monthly_limit > 0) {
    const pct =
      typeof q.usage_percent === 'number'
        ? q.usage_percent
        : (q.used / q.monthly_limit) * 100;
    return `${pct.toFixed(0)}% · ${Math.round(q.used)} / ${Math.round(q.monthly_limit)}`;
  }
  return '--';
}

File: src-tauri/src/models/grok.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Grok CLI OAuth 账号（仅会话账号，不做 API Key）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GrokAccount {
    pub id: String,
    pub email: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub first_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub principal_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub team_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub profile_image_asset_id: Option<String>,

    /// JWT tier 原始值（如 4）
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tier: Option<i64>,
    /// 套餐展示：Free / SuperGrok Lite / SuperGrok / SuperGrok Heavy 等
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plan_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plan_label: Option<String>,

    /// access token（auth.json 字段名为 key）
    pub access_token: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub refresh_token: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub scope: Option<String>,
    /// 过期时间 unix 秒
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<i64>,
    /// auth.json 中 ISO 字符串形式的 expires_at
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expires_at_raw: Option<String>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub oidc_issuer: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub oidc_client_id: Option<String>,
    /// auth.json 顶层 key：官方 CLI 为 `https://auth.x.ai::<client_id>`
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_entry_key: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_mode_raw: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub create_time: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub coding_data_retention_opt_out: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub has_grok_code_access: Option<bool>,

    /// 实时额度（cli-chat-proxy /v1/billing）
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub quota: Option<GrokQuota>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub usage_updated_at: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub token_updated_at: Option<i64>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status_reason: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requires_reauth: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reauth_reason: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub quota_query_last_error: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub quota_query_last_error_at: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub subscription_query_last_success_at: Option<i64>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_raw: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub userinfo_raw: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub billing_raw: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_raw: Option<Value>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub account_note: Option<String>,

    pub created_at: i64,
    pub last_used: i64,
}

/// Grok 额度（对齐 cli-chat-proxy `/v1/billing`）
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct GrokQuota {
    /// 月额度上限（单位与上游 val 一致，通常为美元额度点）
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub monthly_limit: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub used: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub remaining: Option<f64>,
    /// 已用百分比 0-100
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub usage_percent: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub on_demand_cap: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub on_demand_used: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prepaid_balance: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub billing_period_start: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub billing_period_end: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub unlimited_or_free: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub exhausted: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub exhaust_reason: Option<String>,
    /// 便于前端进度条：剩余百分比
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub remaining_percent: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GrokAccountSummary {
    pub id: String,
    pub email: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plan_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tier: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    pub created_at: i64,
    pub last_used: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GrokAccountIndex {
    pub version: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub current_account_id: Option<String>,
    pub accounts: Vec<GrokAccountSummary>,
}

impl GrokAccountIndex {
    pub fn new() -> Self {
        Self {
            version: "1.0".to_string(),
            current_account_id: None,
            accounts: Vec::new(),
        }
    }
}

impl Default for GrokAccountIndex {
    fn default() -> Self {
        Self::new()
    }
}

impl GrokAccount {
    pub fn summary(&self) -> GrokAccountSummary {
        GrokAccountSummary {
            id: self.id.clone(),
            email: self.email.clone(),
            plan_type: self.plan_type.clone(),
            tier: self.tier,
            tags: self.tags.clone(),
            created_at: self.created_at,
            last_used: self.last_used,
        }
    }

    pub fn display_name(&self) -> String {
        if let Some(name) = self.name.as_ref().map(|s| s.trim()).filter(|s| !s.is_empty()) {
            return name.to_string();
        }
        let first = self.first_name.as_deref().unwrap_or("").trim();
        let last = self.last_name.as_deref().unwrap_or("").trim();
        let combined = format!("{} {}", first, last).trim().to_string();
        if !combined.is_empty() {
            return combined;
        }
        self.email.clone()
    }

    pub fn update_last_used(&mut self) {
        self.last_used = chrono::Utc::now().timestamp();
    }
}

/// Device-code OAuth 启动响应
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GrokOAuthStartResponse {
    pub login_id: String,
    pub verification_uri: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_code: Option<String>,
    pub expires_in: u64,
    pub interval_seconds: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub callback_url: Option<String>,
}
